### PR TITLE
Add umask decorator

### PR DESCRIPTION
**Pull request by antevens**

### CODE REVIEW

File: shtdlib.sh [hunk 1]
```diff
@@ -157,6 +157,24 @@ function debug {
     fi
 }
 
+# Umask decorator, changes the umask for a function
+# To use this add a line like the following (without #) as the first line of a function
+# umask_decorator "${FUNCNAME[0]}" "${@}" && return
+function umask_decorator {
+    if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] ; then
+        mask=0007
```
Comment (mroote) [inline shtdlib.sh]: Not sure if you can pass in parameters with the decorators but might be useful to have the mask value configurable.

**Comment by antevens** [comment]
@hrpatel I added configurable parameters to the decorator pattern just using variables, I think we might be able to do the same for system settings no ?

File: shtdlib.sh [hunk 2]
```diff
@@ -157,6 +157,26 @@ function debug {
...
 
+# Umask decorator, changes the umask for a function
+# To use this add a line like the following (without #) as the first line of a function
+# umask_decorator "${FUNCNAME[0]}" "${@}" && return
+# To specify a different umask set the umask_decorator_mask variable to the
+# desired umask.
+function umask_decorator {
+    if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] ; then
+        mask="${umask_decorator_mask:-0007}"
+        original_mask="$(umask)"
```
Comment (hrpatel) [inline shtdlib.sh]: I asked this earlier.. does a subshell have the same `umask` as the current shell?

Comment (antevens) [reply shtdlib.sh]: I thought I already replied to that, umasks are always inherited yes.

File: shtdlib.sh [hunk 3]
```diff
@@ -157,6 +157,26 @@ function debug {
...
 }
 
+# Umask decorator, changes the umask for a function
+# To use this add a line like the following (without #) as the first line of a function
+# umask_decorator "${FUNCNAME[0]}" "${@}" && return
+# To specify a different umask set the umask_decorator_mask variable to the
+# desired umask.
+function umask_decorator {
+    if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] ; then
+        mask="${umask_decorator_mask:-0007}"
```
Comment (mroote) [inline shtdlib.sh]: Should make this a local variable in case we use a `mask` variable elsewhere.

Comment (antevens) [reply shtdlib.sh]: Somehow I thought all variables defined inside functions were local, changed.

File: shtdlib.sh (hunk 2 above)
Comment (hrpatel) [reply shtdlib.sh]: In another PR - which I haven't read yet :P

**Pull request merged**